In [1]:
from IPython.display import HTML
HTML('''
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.
js"></script>
<script>
code_show=true;
function code_toggle() {
if (code_show){
$('div.jp-CodeCell > div.jp-Cell-inputWrapper').hide();
} else {
$('div.jp-CodeCell > div.jp-Cell-inputWrapper').show();
}
code_show = !code_show
}
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" 
value="Click here to toggle on/off the raw code."></form>
''')

In [2]:
from IPython.core.display import HTML
HTML("""
<style>
.output_png {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
}
</style>
""")

<center><img src="./images/Page_Title.png" width="80%" height="80%"></center>

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">Introduction</h1>

Tired of blackboxes in time series deep learning models? This post will discuss a state-of-the-art, hot off-the-press forecasting model called NBEATSx or Neural Basis Expansion Analysis for Time Series with Exogenous Variables. We’ll apply this model in the residential energy consumption context, where we’ll explore how this new model incorporates explainability through time series decomposition and how its performance, in terms of mean absolute percentage error, compares to other existing time series models from basic (e.g., season naive) to more complex (e.g., Prophet).

<h2 style="color:#FFC40F">Motivation</h2>

***

Have you ever experienced bill shock when you received your monthly energy bill? It's common, especially for first-time renters or homeowners who are still familiar with their energy consumption and utility rates. Underestimating energy usage and rates can strain household budgets, causing stress and anxiety. That's where the power of information comes into play. By providing consumers with the tools to make informed decisions about their energy usage, we can empower individuals to take control of their consumption patterns. Whether through real-time monitoring, personalized energy-saving tips, or access to energy usage data, the goal is to arm consumers with the knowledge they need to optimize their energy usage and save money.

<h2 style="color:#FFC40F">Background of the Study</h2>

***

<h3 style="color:#808080">Residential Energy Use-Case</h3>

Residential energy consumption is the amount of energy households use for varied purposes, such as heating, cooling, lighting, and operating appliances and electronic devices. Energy is crucial for maintaining our daily comfort and convenience at home.

However, residential consumers often face challenges when managing their energy usage and the associated costs. One common challenge is the phenomenon known as "bill shock." Bill shock occurs when consumers receive their monthly energy bills with unexpectedly high charges. This is particularly true among first-time renters or homeowners who may underestimate their energy usage or be unaware of the utility rates. Consequently, the unanticipated energy cost increase adversely affects consumers' ability to plan and manage their expenses effectively. 

Consumers have turned to various solutions to alleviate the challenges of bill shock. Some rely on energy-tracking apps to monitor their usage. Still, these apps often require manual input of appliance information or the purchase of expensive smart devices. This manual input can be time-consuming and prone to errors, while the cost of smart devices may be prohibitive for some.

As such, there is a growing need for an automated solution to effortlessly track energy use, predict monthly bills, and maintain a comprehensive record of past usage.

<center><img src="./images/Kardo.png" width="60%" height="60%"></center>
<center>Typical Residential Setup</center>

<h3 style="color:#808080">Time Series Forecasting</h3>

Time series analysis opens up a world of possibilities for forecasting future values based on sequential data. Whether studying residential energy consumption over different time intervals like annual, monthly, daily, or even hourly readings, this analytical approach helps us delve into patterns, spot anomalies, and assess the impact of external factors. Common models like Autoregressive Integrated Moving Averages (ARIMA), Exponential Smoothing (ES), and Seasonal Decomposition of Time Series (STL) capture trends, dependencies, and seasonal patterns in energy usage. These models could help empower us homeowners and energy providers to optimize consumption, make informed decisions, and achieve cost savings.

But what makes it even more exciting is the advent of state-of-the-art models that have revolutionized the accuracy of time series forecasting.

Imagine a world where we can harness the power of deep neural networks to capture non-linear relationships within time series data. That's exactly what MLP-based models offer. By diving deep into the data, these models excel at forecasting short-term dependencies, providing valuable insights into the energy consumption trends of households.

But what about long-term patterns? That's where Recurrent Neural Networks (RNNs) steal the show. By incorporating recurrent connections, RNNs can capture and retain crucial information, helping us understand the long-term dynamics of energy consumption. Variants like LSTM and GRU further enhance the model's ability to grasp the complexity of time series data.

Now, let's talk about transformers. Originally designed for natural language processing, transformers have found their way into the realm of time series analysis. With their self-attention mechanisms, these models can uncover global dependencies, revealing the long-term patterns that influence energy consumption.

But the innovations don't stop there. Temporal Convolutional Networks (TCNs) have emerged as a powerful player in time series analysis. These models leverage dilated convolutions to capture both short and long-range dependencies simultaneously. Their scalable and efficient architecture makes them an excellent choice for analyzing sequential data, such as energy consumption records.

Ensemble methods add another layer of sophistication to time series modeling. By combining predictions from multiple models, we can harness the strengths of each approach and achieve enhanced forecasting accuracy. Techniques like stacking and boosting pave the way for more robust and reliable predictions.

While the accuracy of these state-of-the-art models is impressive, there's another crucial aspect to consider: explainability. In the world of complex models, understanding and interpreting their inner workings becomes paramount. Explainability allows us to unravel the factors driving the predictions, uncover hidden relationships, and gain valuable insights. Regarding decision-making processes, this transparency instills confidence. It empowers stakeholders to make informed choices based on the forecasts generated by these advanced models.

<center><img src="./images/NBEATSx.png" width="80%" height="80%"></center>
<center>NBEATSx Architecture</center>

Time series analysis has come a long way, thanks to the advancements in state-of-the-art models. These powerful tools have revolutionized forecasting accuracy, whether MLP-based models, RNNs, transformers, TCNs, or ensemble methods. Combining their strengths with explainability can unlock the true potential of time series data. Neural Basis Expansion Analysis with exogenous variables (NBEATSx), a groundbreaking architecture, comes in here. NBEATSx incorporates both deep learning and explainability. Recently published by Kin Olivares et al. in the International Journal of Forecasting, this model's beauty lies in its simplicity and effectiveness.[3] At its core is a fully connected layer, one of the simplest neural networks, updated through backcasting to remove learned patterns. The final forecast is obtained by summing up the partial forecasts of the model stacks. Notably, NBEATSx offers interpretability by performing time series decomposition to reveal trends and seasonality patterns.

<h3 style="color:#808080">Dataset</h3>

<center><img src="./images/Dataset.png" width="70%" height="70%"></center>

For our pilot study, as local Philippine data is unavailable, we will use a proxy dataset called the Hourly Usage of Energy (HUE) Dataset for Buildings in British Columbia. This dataset, provided by Simon Fraser University in Canada, covers the period from June 2012 to September 2015. It includes various features such as date, day, month, energy consumption in kilowatt-hours (kWh), hour of the day, humidity levels, holiday indicators, pressure readings, and temperatures. By analyzing this dataset, we aim to gain insights into energy consumption patterns and explore the effectiveness of different time series models in forecasting and optimizing energy usage. Although not specific to the Philippines, this dataset provides a valuable starting point for our pilot study. You can access the dataset here: https://www.nature.com/articles/s41597-022-01455-7 [2]

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">Methodology</h1>

<center><img src="./images/Methodology.png" width="90%" height="90%"></center>

<h2 style="color:#FFC40F">Setting Up</h2>

***

Before everything, we must first install the necessary libraries. We used the following non-standard Python libraries

* `tqdm`, `seaborn`, `scikit-learn`, `torch`, and `neuralforecast`.

In [3]:
# pip install tqdm
# pip install seaborn
# pip install scikit-learn
# pip install torch
# pip install neuralforecast

After installing the non-standard Python libraries, let us import all necessary libraries.

**Basic Imports:** `os`, `itertools`, `tqdm`, `numpy`, `pandas`, `matplotlib`, `seaborn`.

**Deep Learning:** `sklearn`, `torch`, and `neuralforecast`.

In [4]:
# Quick fix for permission error
import os
os.environ['XDG_CACHE_HOME'] = '/home/msds2023/fgarcia/.cache'

In [5]:
# Basic imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import seaborn as sns
import itertools
from tqdm import trange, tqdm

# NBEATSx imports
from neuralforecast.models import NBEATSx
from neuralforecast import NeuralForecast
from neuralforecast.tsdataset import TimeSeriesDataset
from neuralforecast.losses.pytorch import MAPE
from sklearn.metrics import (mean_absolute_error,
                             mean_absolute_percentage_error)

Let us also check if there is an available GPU for Deep Learning.

In [6]:
import torch
device = (torch.device('cuda') if torch.cuda.is_available()
          else torch.device('cpu'))
print(f"Your Device is {device}.")

Your Device is cuda.


<h2 style="color:#FFC40F">Data Preprocessing</h2>

***

The Hourly Usage of Energy (HUE) dataset comprises multiple csv files, including Residential 1-28, Solar, Weather WYJ, Weather YVR, and Holidays. However, we will only work with three CSV files for our specific analysis: holidays, weatherYVR, and residential_1.

The holidays CSV file contains information about dates, days of the week, weekends, holidays, and daylight saving time (dst). The weather YVR CSV file includes data on dates, hours, temperature, humidity, pressure, and weather conditions. Lastly, the residential_1 CSV file details dates, hours, and energy consumption in kilowatt-hours (kWh).

<h3 style="color:#808080">Energy Consumption</h3>

In this part of the post, we pre-processed the data from residential_1.csv which will be used as the basis of residential energy consumption in the succeeding sections. The following pre-processing were were conducted:

1. *Converted the date column to datetime format.* This is to ensure that the dates are uniformly represented in the dataset. This allowed us to access a wide range of date-specific functionalities and operations, and enabled us to facilitate easy and efficient time-based indexing which is necessary in performing time series analysis.
2. *Drop duplicates.* Dropping duplicates allowed us to preserve the integrity of the dataset by ensuring that each data point is a unique and distinct observation. This also helped us avoid any bias and removed any unnecessary consumption of storage space. 
3. *Remove the first day.* Due to incomplete data, where the first day's data point started at 1 AM instead of 12 midnight, we made the decision to exclude the first data point.
4. *Set datetime to index*. Since we did step 1, we can now set the datetime column to index which can allow us to take advantage of varied built-in time-series models in python (e.g Pandas).
5. *Fill in missing dates and backfill to impute.* We did this to ensure continuity in data and each data point is evenly spaced. The method we used to impute the values is backfill.
6. *Resample to day.* In this step we resampled the hourly data to a daily perspective for our time-series analysis. This allowed us to aggregate the energy consumption data at a daily level.

In [7]:
# Load dataset
df = pd.read_csv('Residential_1.csv')

In [8]:
# Preprocessing
df_data = df.copy()

# Convert the date column to datetime format
df_data['date'] = pd.to_datetime(df_data['date'])

# Create the datetime column by adding the hour to the date
df_data['datetime'] = df_data['date'] + pd.to_timedelta(df_data['hour'],
                                                        unit='H')
df_data = df_data.drop(columns=['date', 'hour'])

# Drop duplicates
df_data = df_data.drop_duplicates(subset='datetime')
df_data = df_data.sort_values(by='datetime')

# Remove the first day and last day (not complete)
df_data = df_data[23:-1]

# Set datetime to index
df_data = df_data.set_index('datetime')
df_data.index = pd.to_datetime(df_data.index)

# Fill in missing dates
datetime_range = pd.date_range(start='2012-06-02 00:00:00',
                               end='2015-10-02 23:00:00', freq='H')
df_date = pd.DataFrame(index=datetime_range)
df_comp = df_date.merge(df_data, how='outer',
                        left_index=True, right_index=True)

# Backfill to impute
df_energy = df_comp.fillna(method='bfill')

# Resample to day
df_day = df_energy[['energy_kWh']].resample('D').sum()[:-2]

<h3 style="color:#808080">Exogeneous Variables</h3>

As mentioned above, we used the Weather_YVR.csv and Holiday.csv as our exogenous variables.

For the Weather_YVR.csv, we only used the  date, hour, temperature, humidity, and pressure features. We conducted the following pre-processing steps:

1. *Clean up the datetime column.* In this part of the code, we cleaned up and transformed the date and hour columns into a single datetime column in the dataframe *df_exo*.
2. *Drop duplicates.* We dropped any duplicate rows based on the 'datetime' column in the dataframe df_exo and sorted the dataframe in chronological order.
3. *Set datetime to index.* This step allowed us to access and process the data using built-in time-series models in python. 
4. *Resample to day.* We did this to ensure continuity in data and each data point is evenly spaced. The method we used to impute the values is backfill. 
5. *Add more feature to capture long term seasonality.* This step is crucial as it allowed us to incorporate long-term seasonality into the analysis, the 'dayofweek' and 'month' features are added to the 'df_exo' dataframe. These features captured the day of the week (0-6, where Monday is 0 and Sunday is 6) and the month (1-12) respectively, based on the index values of the dataframe.
6. *Dataframe containing both consumption and exogeneous.* This is the step where we merged the above previous dataframe with the current dataframe. The resulting dataframe, df_all, contains both the energy consumption data and the exogenous features.

In [9]:
# Load data
df_exo = pd.read_csv('Weather_YVR.csv',
                     usecols=['date', 'hour', 'temperature',
                              'humidity', 'pressure'])

# Clean up datetime
df_exo['date'] = pd.to_datetime(df_exo['date'])
df_exo['datetime'] = df_exo['date'] + pd.to_timedelta(df_exo['hour'],
                                                      unit='H')
df_exo = df_exo.drop(columns=['date', 'hour'])

# Drop duplicates
df_exo = df_exo.drop_duplicates(subset='datetime')
df_exo = df_exo.sort_values(by='datetime')

# Set datetime to index
df_exo = df_exo.set_index('datetime')
df_exo.index = pd.to_datetime(df_exo.index)

# Resample to day
df_exo = df_exo[['temperature', 'humidity', 'pressure']].resample('D').mean()

# Add more feature to capture long term seasonality
df_exo['dayofweek'] = df_exo.index.dayofweek
df_exo['month'] = df_exo.index.month

# Dataframe containing both consumption and exogeneous
df_all = df_exo.merge(df_day, left_index=True, right_index=True)

For the Holiday.csv we only used the date and holiday columns as our features. We didn’t merge this with df_all as we used this for the Prophet time-series model. 

In [10]:
# Load holiday exogeneous variable
df_holiday = pd.read_csv('Holidays.csv', usecols=['date', 'holiday'])
df_holiday['ds'] = pd.to_datetime(df_holiday['date'])
df_holiday = df_holiday.dropna().drop(columns='date')

The final part Dataframe for NBEATSx is based on the specific format that NBEATSx accepts.

In [11]:
# Dataframe for NBEATSx
df_beats = df_all.copy()
df_beats['unique_id'] = 1
df_beats = df_beats.reset_index()
df_beats = df_beats.rename(columns={'energy_kWh': 'y', 'index': 'ds'})
df_beats = df_beats.reset_index()

<h2 style="color:#FFC40F">Exploratory Data Analysis</h2>

***

In our analysis, we utilized time series plots to examine the overall trend of energy consumption in British Columbia from June 2012 to September 2015. The first plot focused on daily energy consumption, providing insights into trend and seasonality. The second plot incorporated exogenous variables (temperature, humidity, and atmospheric pressure) alongside energy consumption to explore their relationships.

<h3 style="color:#808080">Energy Consumption Time Series Plot</h3>

In [12]:
df_hday = (df_holiday.merge(df_day.reset_index(),
                            how='left',
                            left_on='ds',
                            right_on='index')
           [['ds', 'energy_kWh']]
           .dropna())

datetime_range_eda = pd.date_range(start='2012-06-02 00:00:00',
                                   end='2015-10-02 23:00:00',
                                   periods=11)

# Plot whole dataset
fig, ax = plt.subplots(1, 1, figsize=(20, 5))
sns.lineplot(data=df_day, y=df_day.energy_kWh, x=df_day.index,
             color='orange', legend=None)
plt.xticks(datetime_range_eda)
ax.xaxis.set_major_formatter(mdates.DateFormatter('%b %Y'))

# Plot
plt.ylabel("Energy(kWh)")
plt.xlabel("Day")
plt.title('Daily Time Series Plot of Energy Consumption',
          fontsize=20, pad=20)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Daily_Time_Series.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Daily_Time_Series.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:100%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In the plot above, we can see that energy consumption increases each year in the middle of October and decreases by the end of February. The energy in the months with high consumption typically ranges from 30 to 36kWh per day, but the consumption could still go up to 41 to 47kWh. While in the months with low consumption, the energy typically ranges from 20 to 25 kWh per day. Below we could explore further why this might be the case. But it is important to keep in mind that this dataset is taken from British Columbia.

<h3 style="color:#808080">Exogeneous Variables Time Series Plot</h3>

In [13]:
# Plot exogeneous variable
df_eda = df_all[:'2015-08-31'].reset_index()
df_eda['year'] = pd.DatetimeIndex(df_eda['index']).year
df_line = (df_eda.groupby(['year', 'month'])['temperature', 'humidity',
                                             'pressure', 'energy_kWh']
           .mean().reset_index())
df_line['date'] = pd.to_datetime(df_line[['year', 'month']].assign(DAY=1))
df_line2 = df_line.drop(columns=['year', 'month'])

datetime_range_eda = pd.date_range(start='2012-06-02 00:00:00',
                                   end='2015-10-02 23:00:00',
                                   periods=11)

fig, ax = plt.subplots(figsize=(20, 6))

sns.lineplot(x='date', y='value', hue='variable',
             data=pd.melt(df_line2, 'date'),
             palette='rocket', legend=None)

plt.ylabel("Value")
plt.xlabel("Month-Year")
plt.title('Time Series Plot of Energy Consumption and Exogenous Variables',
          fontsize=20, pad=20)
ax.text(pd.Timestamp("2015-06-25"), 10, 'Temperature(°C)', color='#321854')
ax.text(pd.Timestamp("2015-06-25"), 32, 'Energy(kWh)', color='#cc5f16')
ax.text(pd.Timestamp("2015-06-25"), 65, 'Humidity(%)', color='#a81b64')
ax.text(pd.Timestamp("2015-06-25"), 95, 'Pressure(kPa)', color='red')
ax.xaxis.set_major_formatter(mdates.DateFormatter('%b %Y'))
plt.xticks(datetime_range_eda)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Exo_Time_Series.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Exo_Time_Series.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:100%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In the plot above, first, let's focus on the relationship between energy consumption `Energy(kWh)` and `Temperature`. Same with the previous plot, we can see that starting from the middle of October, energy consumption starts to increase, and during this time, temperature decreases. On the other hand, starting from March, energy consumption decreases and temperature increases. This could be attributed to the dataset being based on British Columbia, where there are four seasons. 

Based on the illustration below, we could say that temperature decreases during the latter months of the Fall and continues until the end of the Winter season. Then temperature increases at the start of the Summer and continues until the Spring or early Fall seasons. This means that during the months with low temperatures, people consume more Energy because of their use of heaters to fight against the cold weather. It follows that during the months with high temperatures, they consume less Energy because of less use of their heaters due to the hotter weather.

Looking at `Humidity`, we see it follows the `Energy(kWh)` trend. However, it is misleading to think that high humidity directly affects energy consumption. It is more proper to relate humidity and temperature, which are inversely related. This means that the relative humidity is higher if the air is cooler or if the temperature is lower.


<center><img src="./images/Seasons.png" width="70%" height="70%"></center>
<center>Seasons in British Columbia, Canda</center>

<h2 style="color:#FFC40F">Time Series Forecasting (NBEATSx)</h2>

***

We predict the energy consumption of Kardo by showcasing the state-of-the-art forecasting model, NBEATSx. To fully utilize the model's capability, we aggregated the daily energy consumption with the exogenous variables collected. The forecasting will be done in 7-day (week) and 30-day (month) ahead forecasts to cover both short-term and long-term areas. To evaluate the model's effectiveness in predicting energy consumption, the results of NBEATSx will be compared to Seasonal Naive, which will serve as the baseline model. The evaluation metric is set to Mean Absolute Percentage Error (MAPE) to easily contextualize the magnitude of the scores when compared with other scores. Lastly, the models below are already hypertuned to produce the optimal MAPE.

<h3 style="color:#808080">Short-Term (7-day ahead) Forecast</h3>

Before proceeding with the model, we must split the dataset into train and test splits. This is the accuracy of the models for data that it has not learned yet. We set the test set to be the last 7 days of the dataset's time period for this short-term forecast.

Next, we train the model using the train set and iterate the training process in 1000 epochs. We optimize the following parameters to get the lowest MAPE:

- input size, which is the size of the autoregressive input where the backcast is based
- futr_exog_list, which is the list of future known variables
- scaler_type, which is the type of scaler for temporal inputs normalization
- n_harmonics sets the number of harmonic oscillations factored in the seasonality stack
- n_polynomials set the degree of the trend stack
- stack_types indicates which stacks the model will use
- activation sets the activation function used in the neural network
- learning_rate indicates how much the weights will change per iteration
- mlp_units indicates the structure of hidden layers for each stack type
- num_lr_decays indicates the number of times the learning rate is decayed or reduced in the training

The random_seed is set so we can reproduce the results. Then, we use the model on the test set to evaluate how well the model performs on the energy consumption dataset and look at the MAPE.


In [14]:
# Set forecast horizon
hor_7 = 7

# Splitting the dataset
df_beats_train_7 = df_beats[:-hor_7]
df_beats_test_7 = df_beats.iloc[-hor_7:].drop(columns='y')

# 7 day-ahead forecast
models = [NBEATSx(h=hor_7,
                  input_size=(4*hor_7),
                  max_steps=1000,
                  futr_exog_list=['temperature', 'humidity',
                                  'pressure', 'dayofweek', 'month'],
                  scaler_type='robust',
                  n_harmonics=3,
                  n_polynomials=3,
                  stack_types=['identity', 'trend', 'seasonality'],
                  activation='LeakyReLU',
                  loss=MAPE(),
                  learning_rate=1e-4,
                  mlp_units=3*[[256, 256]],
                  num_lr_decays=1,
                  random_seed=630
                  )]
nf = NeuralForecast(models=models, freq='D')

# Train & Predict
nf.fit(df=df_beats_train_7)
y_pred_df = nf.predict(futr_df=df_beats_test_7)

# Save to dataframe
df_nbtx_res_7 = y_pred_df.set_index('ds')

Global seed set to 630
2023-06-19 14:26:33.547963: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-06-19 14:26:33.636288: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.


Sanity Checking: 0it [00:00, ?it/s]

Training: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Predicting: 0it [00:00, ?it/s]

For the baseline model, Seasonal Naive, we use the previous week's energy consumption as the forecast for the succeeding week. Then we project the results as a dataframe, bar graph, and trend line showing the true values with the prediction of the two models.

In [15]:
# Baseline
df_snaive_7 = df_day.copy()
df_snaive_7['Seasonal_Naive'] = df_day[['energy_kWh']].shift(hor_7)
df_snaive_7 = df_snaive_7[['Seasonal_Naive']].iloc[-hor_7:]

# Combine all predictions and actual
df_predict_7 = df_day.copy()
df_predict_7 = df_predict_7.rename(columns={'energy_kWh': 'Actual'})
df_predict_7 = df_predict_7[-hor_7:]
df_predict_7 = pd.concat([df_predict_7, df_snaive_7, df_nbtx_res_7], axis=1)

# Append the model score to the results dataframe
model_ls = []
mae_ls = []
mape_ls = []
for col in df_predict_7.columns[1:]:
    model_ls.append(col)
    mae_ls.append(mean_absolute_error(df_predict_7['Actual'],
                                      df_predict_7[col]))
    mape_ls.append(mean_absolute_percentage_error(df_predict_7['Actual'],
                                                  df_predict_7[col]))

df_results_7 = pd.DataFrame(
    {'Model': model_ls, 'MAE': mae_ls, 'MAPE': mape_ls})
display(df_results_7)

Model       MAE      MAPE
0  Seasonal_Naive  3.041286  0.112502
1         NBEATSx  1.247066  0.045908

In [16]:
# Plot Results of Different Models
fig, ax = plt.subplots(figsize=(8, 3))
ax.barh(df_results_7['Model'], df_results_7['MAPE'], color='#FAA01D')
ax.set_title('MAPE Comparison of Seasonal Naive and NBEATSx')
ax.set_xlabel('Mean Absolute Percentage Error')
ax.set_yticklabels(['Seasonal Naive', 'NBEATSx'])
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Short_MAPE_1.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Short_MAPE_1.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In [17]:
# Time Series Prediction
fig, ax = plt.subplots(figsize=(15, 5))
ax.plot(df_predict_7.index, df_predict_7['Actual'],
        label='Actual', color='cornflowerblue', linewidth=3)
ax.plot(df_predict_7.index, df_predict_7['Seasonal_Naive'],
        label='Seasonal Naive', linestyle='--', color='indianred', linewidth=3)
ax.plot(df_predict_7.index, df_predict_7['NBEATSx'],
        label='NBEATSx', linestyle='--', color='#FAA01D', linewidth=3)
ax.set_title('Actual vs. Seasonal Naive & NBEATSx Short-Term Prediction')
ax.set_ylabel('Daily Energy Consumption (kWh)')
ax.set_xlabel('Test Prediction Dates')
ax.legend()
ax.set_ylim(20, 35)
ax.tick_params(axis='x', rotation=15)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Short_Prediction_1.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Short_Prediction_1.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

Looking at the results of the 7-day ahead forecast, NBEATSx beats out Seasonal Naive by far (4.6% to 11.3%). The visualization of the predictions and the actual value shows that NBEATSx could follow the trend of the actual energy consumption (although the trend changes are minimal in the model) compared to the baseline model.

<h3 style="color:#808080">Long-Term (30-day ahead) Forecast</h3>

The long-term forecast for energy consumption has almost the same flow as the short-term forecast. There are only two differences in the method: the forecast horizon, which is set to 30, and the hypertuned parameters of the models. The updated parameters for the month-ahead forecast are the following:
- input_size set to a longer period to factor in the longer time horizon
- n_harmonics and n_polynomials since the trend and seasonality are changed
- learning_rate set to a smaller scale due to different granularity

In [18]:
# Set forecast horizon
hor_30 = 30

# Splitting the dataset
df_beats_train_30 = df_beats[:-hor_30]
df_beats_test_30 = df_beats.iloc[-hor_30:].drop(columns='y')

# 30 day-ahead forecast
models = [NBEATSx(h=hor_30,
                  input_size=135,  # 7-day lookback
                  max_steps=1000,
                  futr_exog_list=['temperature', 'humidity',
                                  'pressure', 'dayofweek', 'month'],
                  scaler_type='robust',
                  n_harmonics=2,
                  n_polynomials=1,
                  stack_types=['identity', 'trend', 'seasonality'],
                  activation='LeakyReLU',
                  loss=MAPE(),
                  learning_rate=1e-5,
                  mlp_units=3*[[256, 256]],
                  num_lr_decays=1,
                  random_seed=630
                  )]
nf = NeuralForecast(models=models, freq='D')

Global seed set to 630


In [19]:
# Train & Predict
nf.fit(df=df_beats_train_30)
y_pred_df = nf.predict(futr_df=df_beats_test_30)

# Save to dataframe
df_nbtx_res_30 = y_pred_df.set_index('ds')

Sanity Checking: 0it [00:00, ?it/s]

Training: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Predicting: 0it [00:00, ?it/s]

The baseline model is set to the previous month's energy consumption.

In [20]:
# Baseline
df_snaive_30 = df_day.copy()
df_snaive_30['Seasonal_Naive'] = df_day[['energy_kWh']].shift(hor_30)
df_snaive_30 = df_snaive_30[['Seasonal_Naive']].iloc[-hor_30:]

In [21]:
# Combine all predictions and actual
df_predict_30 = df_day.copy()
df_predict_30 = df_predict_30.rename(columns={'energy_kWh': 'Actual'})
df_predict_30 = df_predict_30[-hor_30:]
df_predict_30 = pd.concat(
    [df_predict_30, df_snaive_30, df_nbtx_res_30], axis=1)

# Append the model score to the results dataframe
model_ls = []
mae_ls = []
mape_ls = []
for col in df_predict_30.columns[1:]:
    model_ls.append(col)
    mae_ls.append(mean_absolute_error(df_predict_30['Actual'],
                                      df_predict_30[col]))
    mape_ls.append(mean_absolute_percentage_error(df_predict_30['Actual'],
                                                  df_predict_30[col]))

df_results_30 = pd.DataFrame(
    {'Model': model_ls, 'MAE': mae_ls, 'MAPE': mape_ls})
display(df_results_30)

Model       MAE      MAPE
0  Seasonal_Naive  4.499367  0.159477
1         NBEATSx  2.562424  0.088149

In [22]:
# Plot Results of Different Models
fig, ax = plt.subplots(figsize=(8, 3))
ax.barh(df_results_30['Model'], df_results_30['MAPE'], color='#FAA01D')
ax.set_title('MAPE Comparison of Seasonal Naive and NBEATSx')
ax.set_xlabel('Mean Absolute Percentage Error')
ax.set_yticklabels(['Seasonal Naive', 'NBEATSx'])
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Long_MAPE_1.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Long_MAPE_1.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In [23]:
# Time Series Prediction
fig, ax = plt.subplots(figsize=(15, 5))
ax.plot(df_predict_30.index, df_predict_30['Actual'],
        label='Actual', color='cornflowerblue', linewidth=3)
ax.plot(df_predict_30.index, df_predict_30['Seasonal_Naive'],
        label='Seasonal Naive', linestyle='--', color='indianred',
        linewidth=3)
ax.plot(df_predict_30.index, df_predict_30['NBEATSx'],
        label='NBEATSx', linestyle='--', color='#FAA01D', linewidth=3)
ax.set_title('Actual vs. Seasonal Naive & NBEATSx Long-Term Prediction')
ax.set_ylabel('Daily Energy Consumption (kWh)')
ax.set_xlabel('Test Prediction Dates')
ax.legend()
ax.set_ylim(5, 40)
ax.tick_params(axis='x', rotation=15)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Long_Prediction_1.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Long_Prediction_1.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The results of the 30-day ahead forecast show NBEATSx having a better MAPE than Seasonal Naive (8.8% vs 15.9%). It is expected to have a higher MAPE for the month ahead forecast since we forecast longer time horizons, leading to less accuracy. Another note is that the trend is no longer correctly captured in NBEATSx compared to the 7-day ahead forecast. Nonetheless, the model could predict energy consumption with lower MAPE.

<h2 style="color:#FFC40F">XAI (Time Series Decomposition)</h2>

***

One of the features of NBEATSx is its built-in explainability feature. The time series model can be decomposed according to the stacks set in the parameters. These stacks can be the trend, seasonality, and exogenous stacks. NBEATSx computes a level prediction that is the same throughout the forecast horizon, then adds the outputs of the different stacks for each forecast. The outputs of each stack will then be the explainability of the model as a decompose function is available to pull out all the outputs of each stack. This way, we can visualize how each forecast is affected by the trend, seasonality, and exogenous variables.

However, one current limitation of the NBEATSx is that the exogenous stack is unavailable when the time series is decomposed. Only the trend and seasonality stacks are available and will be shown in this post. This results in an altered version of the forecasting model as exogenous variables are left, and the mlp_units will be updated to only the two stacks.

In [24]:
# Dataframe for NBEATSx time decomposition
df_beatsxai = df_day.reset_index()
df_beatsxai = df_beatsxai.rename(columns={'index': 'ds', 'energy_kWh': 'y'})
df_beatsxai['unique_id'] = 1
df_beatsxai = df_beatsxai.reset_index()
display(df_beatsxai)

index         ds       y  unique_id
0         0 2012-06-02  20.331          1
1         1 2012-06-03  22.844          1
2         2 2012-06-04  25.610          1
3         3 2012-06-05  24.127          1
4         4 2012-06-06  27.718          1
...     ...        ...     ...        ...
1211   1211 2015-09-26  31.067          1
1212   1212 2015-09-27  29.411          1
1213   1213 2015-09-28  26.998          1
1214   1214 2015-09-29  26.975          1
1215   1215 2015-09-30  23.496          1

[1216 rows x 4 columns]

<h3 style="color:#808080">Short-Term Time Series Decomposition</h3>

In [25]:
# 7 day-ahead daily forecast
horizon = 7
models = [NBEATSx(h=horizon,
                  input_size=(4*horizon),  # 7-day lookback
                  max_steps=1000,
                  n_harmonics=2,
                  n_polynomials=3,
                  stack_types=['trend', 'seasonality'],
                  activation='LeakyReLU',
                  loss=MAPE(),
                  learning_rate=1e-5,
                  mlp_units=2*[[256, 256]],
                  num_lr_decays=1,
                  random_seed=630
                  )]
nf = NeuralForecast(models=models, freq='D')

nf.fit(df=df_beats_train_7)
y_pred_df = nf.predict(futr_df=df_beats_test_7)

Global seed set to 630


Sanity Checking: 0it [00:00, ?it/s]

Training: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Predicting: 0it [00:00, ?it/s]

In [26]:
# NBEATSx decomposition plot
model = nf.models[0]
dataset, *_ = TimeSeriesDataset.from_df(df=df_beatsxai)
y_hat = model.decompose(dataset=dataset, random_seed=630)

Predicting: 0it [00:00, ?it/s]

The output of the decompose function results in a 1 x 3 x n matrix where n is the forecast horizon. The size of three corresponds to the level or baseline stack, trend, and seasonality in that order. Adding the outputs of the three stacks should result in the prediction output.

In [27]:
# Plot
fig, ax = plt.subplots(3, 1, figsize=(12, 10), sharex=True)

ax[0].plot(df_day[-hor_7:].values, label='Actual',
           color='cornflowerblue', linewidth=3)
ax[0].plot(y_hat.sum(axis=1).flatten(), label='Forecast',
           color="#FAA01D", linestyle='--', linewidth=3)
ax[0].set_title('Actual vs. Forecast')
ax[0].set_ylabel('Energy Consumption (kWh)')
ax[0].legend()

# Trend
ax[1].plot(y_hat[0, 1]+y_hat[0, 0], color="#FAA01D", linewidth=3)
ax[1].set_title('Trend Decomposition')
ax[1].set_ylabel('Energy Consumption (kWh)')

# Seasonality
ax[2].plot(y_hat[0, 2], color="#FAA01D", linewidth=3)
ax[2].set_title('Seasonality Decomposition')
ax[2].set_ylabel('Energy Consumption (kWh)')
ax[2].set_xlabel('Day of the Week')
custom_labels = [0, 'Thu', 'Fri', 'Sat', 'Sun', 'Mon', 'Tue', 'Wed']
ax[2].set_xticklabels(custom_labels)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Short_Decomp.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Short_Decomp.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:70%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

Looking at the explainability of the 7-day ahead forecast, the trend increases exponentially from Thursday to Wednesday. This suggests the increasing trend of energy consumption of Kardo as days go by. The seasonality plot indicates that the Kardo has peak consumption on Saturdays and floor consumption on Thursdays.

Note that we added the baseline prediction with the trend to show that adding the trend and seasonality will result in the prediction (since the level prediction is just a horizontal line when plotted).

<h3 style="color:#808080">Long-Term Time Series Decomposition</h3>

The month-ahead forecast has the flow as the week-ahead forecast with only the forecast horizon and some parameters being updated to adjust with the longer forecast horizon. The outputs will also be the same, where the trend and seasonality are the basis for explainability.

In [28]:
# 30 day-ahead daily forecast
models = [NBEATSx(h=hor_30,
                  input_size=(4*hor_30),
                  max_steps=1000,
                  n_harmonics=2,
                  n_polynomials=4,
                  stack_types=['trend', 'seasonality'],
                  activation='LeakyReLU',
                  loss=MAPE(),
                  learning_rate=1e-5,
                  mlp_units=2*[[256, 256]],
                  num_lr_decays=1,
                  random_seed=630
                  )]
nf = NeuralForecast(models=models, freq='D')

# Train & Predict
nf.fit(df=df_beats_train_30)
y_pred_df = nf.predict(futr_df=df_beats_test_30)

Global seed set to 630


Sanity Checking: 0it [00:00, ?it/s]

Training: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Predicting: 0it [00:00, ?it/s]

In [29]:
# NBEATSx decomposition plot
model = nf.models[0]
dataset, *_ = TimeSeriesDataset.from_df(df=df_beatsxai)
y_hat = model.decompose(dataset=dataset, random_seed=630)

Predicting: 0it [00:00, ?it/s]

In [30]:
# Plot
fig, ax = plt.subplots(3, 1, figsize=(12, 10), sharex=True)

ax[0].plot(df_day[-hor_30:].values, label='Actual',
           color='cornflowerblue', linewidth=3)
ax[0].plot(y_hat.sum(axis=1).flatten(), label='Forecast',
           color="#FAA01D", linestyle='--', linewidth=3)
ax[0].set_title('Actual vs. Forecast')
ax[0].set_ylabel('Energy Consumption (kWh)')
ax[0].legend()

# Trend
ax[1].plot(y_hat[0, 1]+y_hat[0, 0], color="#FAA01D", linewidth=3)
ax[1].set_title('Trend Decomposition')
ax[1].set_ylabel('Energy Consumption (kWh)')

# Seasonality
ax[2].plot(y_hat[0, 2], color="#FAA01D", linewidth=3)
ax[2].set_title('Seasonality Decomposition')
ax[2].set_ylabel('Energy Consumption (kWh)')
ax[2].set_xlabel('Day of the Week')
ax[2].set_xticks(np.arange(30))
custom_labels = ['Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun',
                 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun',
                 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun',
                 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun',
                 'Mon', 'Tue', 'Wed']
ax[2].set_xticklabels(custom_labels)
ax[2].axvline(x=6, color='indianred', linestyle='--')
ax[2].axvline(x=13, color='indianred', linestyle='--')
ax[2].axvline(x=18, color='indianred', linestyle='--')
ax[2].axvline(x=25, color='indianred', linestyle='--')
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Long_Decomp.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Long_Decomp.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:70%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The model indicates that Kardo has an increasing trend from the start of the month until the middle of the month before it starts to decrease until the end. For the seasonality, the forecast shows less consumption from Saturday to Monday, which might indicate less activity in residence. 

Another thing to look at is the magnitude of the stacks' outputs and how it differs for the two forecast horizons. For the short-term forecast, the seasonality has a higher magnitude than the trend stack (subtracting the level prediction). While for long-term forecasts, the trend stack's output is higher in magnitude compared to the output of the seasonality stack.

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">Results and Discussion</h1>

<h2 style="color:#FFC40F">Forecast Comparison</h2>

***

In this section, we will compare the results of NBEATSx to other forecasting models. Also, we will use another explainability method, such as SHAP, to explain the effect of temperature on energy consumption.

<center><img src="./images/Short_MAPE_2.png" width="60%" height="60%"></center>

In short-term forecasting, NBEATSx beats all other models, including seasonal naïve, random forest, and prophet, by having a MAPE of 4.59%. 

<center><img src="./images/Long_MAPE_2.png" width="60%" height="60%"></center>

In long-term forecasting, NBEATSx still beats all other models with a MAPE of 8.81%. As we increase the forecast horizon, it is expected that the accuracy of the forecasting model will drop. Compared to other models in this long-term forecast horizon, it is still one of the best.

Stacking multiple blocks of fully connected layers while efficiently using backcasting and simple MLP allows NBEATSx to forecast more accurately. But the advantage of NBEATSx over other models is its ability to be interpretable, as demonstrated by earlier time series decomposition. Even though the MAPE differs slightly, this advantage makes NBEATSx a more desirable forecasting method.

<h2 style="color:#FFC40F">Other XAI (SHAP)</h2>

***

The results of random forest are interpreted using an explainability method called SHAP. SHAP or Shapley
Additive exPlanations assign feature importance by quantifying the contribution of each
feature to the model prediction. This provides insight into what features would clearly
distinguish between the two classes. The global feature importance plot of the model is shown
below.

<center><img src="./images/SHAP.png" width="60%" height="60%"></center>

The top feature in predicting energy consumption is temperature, humidity, and previous lag value. As the descriptive analysis corroborates, energy consumption decreases as temperature increases. We will explain a single forecast on our test set to further investigate this generalization.

<center><img src="./images/Waterfall.png" width="60%" height="60%"></center>

In this plot, we can see coming from the global average of energy consumption, the high-temperature value (relative to other time of the year) push the consumption towards lower energy consumption. Let us determine how big of a change in energy consumption each temperature increase.

Here we calculated that for every one degrees Celsius increase in temperature, that is an equivalent 0.289 decrease in energy consumption (kWh).

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">Key Takeaways</h1>

Deep Learning methods have proven to outperform classical statistical and machine learning methods in time series forecasting. However, accuracy alone is not the sole consideration. Explainability is equally important to establish trust with users. Fortunately, with NBEATSx, we can achieve both accuracy and explainability in our forecasts.

NBEATSx provides faster and more efficient time series forecasting owing to its MLP-based architecture. In addition, it provides highly accurate forecasts. However, what really sets NBEATSx apart is its unique ability to decompose time series, revealing the underlying trends and seasonality that influence the forecasted values. This enhances the usability and practicality of the model, making it an attractive choice for time series analysis.

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">Github Link</h1>

You may refer also to this github repo for the complete code and data: https://github.com/flippygarcia/msds2023_ml3_individual_project.git

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">References</h1>

[1] Tilford, A. (2023, Mar 30). *Apps to Help You Track Energy Use.* Retrieved June 18, 2023 from https://www.saveonenergy.com/resources/apps-track-energy-use/

[2] Zheng, X. et al. (2022, Jun 22). *A multi-scale time-series dataset with benchmark for machine learning in decarbonized energy grids*. Retrieved June 18, 2023 from https://www.nature.com/articles/s41597-022-01455-7

[3] Olivares, K. et al. (2023, Apr 04). *Neural basis expansion analysis with exogenous variables: Forecasting electricity prices with NBEATSx.* Retrieved June 18, 2023 from https://arxiv.org/pdf/2104.05522.pdf

[4] US Department of Commerce, N. (n.d.). *Discussion on Humidity*. Retrieved June 19, 2023 from Www.weather.gov. https://www.weather.gov/lmk/humidity#:~:text=Relative%20humidity%20(RH)%20(expressed,air%20at%20its%20current%20temperature.)

[5] OpenAI. (2023). *ChatGPT* (May 24 version) [Large language model]

<h1 style="background-color:#FAA01D; color:#F5F5F1; padding: 10px;">Appendix</h1>

In this section, the entire code used for the project is shown. This now includes the time series forecasting of random forest, and prophet. As well as XAI using SHAP and counterfactual.

<h2 style="color:#FFC40F">Additional Exploratory Data Analysis</h2>

***

An autocorrelation plot, or an ACF (Autocorrelation Function) plot, is a graphical tool used in time series analysis to identify any patterns or dependencies within the data. It shows the correlation between a time series and its lagged values. Statsmodel has a built-in function to produce an ACF plot given a time series dataset.

In [31]:
from statsmodels.graphics.tsaplots import plot_acf

In [32]:
# Autocorrelation plot of energy consumption
fig, ax = plt.subplots(figsize=(12, 5))
plot_acf(df_eda[['energy_kWh']], zero=False, ax=ax)
plt.title('Autocorrelation Plot of Energy Consumption',
          fontsize=15, pad=20)
plt.ylabel('Autocorrelation Values')
plt.xlabel('Time Lag')
plt.ylim(-0.2, 1)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Auto_Cor.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Auto_Cor.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The ACF plot indicates a high correlation for all the lagged values from the first to the 30th. The peak correlation occurs at the first lagged values; the lowest is the 30th lagged value. For the trend, it has a seven-day cycle where the trend is decreasing.

The succeeding plots are the histograms of the daily energy consumption and the exogenous variables. This can highlight the behavior of Kardo in terms of energy consumption and the environment he is in.

In [33]:
fig, ax = plt.subplots(figsize=(12, 5))
sns.histplot(df_eda[['energy_kWh']], x='energy_kWh', color='#cc5f16',
             edgecolor='black', bins=50, legend=None,
             kde=True)

plt.ylabel("Count")
plt.xlabel("Energy(kWh)")
plt.title('Histogram of Energy Consumption of Residential Buildings',
          fontsize=15, pad=20)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Histogram_Energy.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Histogram_Energy.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The histogram of the daily energy consumption shows that the energy consumption of Kardo is around 20-30 kWh. The peak energy consumption is around 50 kWh, and there are instances of below 10 kWh, which can result from power interruptions during that day.

In [34]:
fig, ax = plt.subplots(figsize=(12, 5))
sns.histplot(df_eda[['temperature']], x='temperature', color='#321854',
             edgecolor='black', bins=50, legend=None,
             kde=True)

plt.ylabel("Count")
plt.xlabel("Temperature(°C)")
plt.title('Histogram of Temperature of Residential Buildings',
          fontsize=20, pad=20)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Histogram_Temp.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Histogram_Temp.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The temperature histogram shows the variations in the weather in British Columbia, Canada. There are two peak counts, around 5 to 8 degrees Celsius and another at 15 to 19 degrees Celsius, due to the location having four seasons. The range of values is from -5 to around 25 degrees Celsius which extreme values have rare occurrences.

In [35]:
fig, ax = plt.subplots(figsize=(12, 5))
sns.histplot(df_eda[['humidity']], x='humidity', color='#a81b64', bins=50, legend=None,
             kde=True, palette='mako')

plt.ylabel("Count")
plt.xlabel("Humidity (%)")
plt.title('Histogram of Humidity of Residential Buildings',
          fontsize=20, pad=20)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Histogram_Humidity.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Histogram_Humidity.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The humidity in British Columbia plays around 70% to 90% humidity during the time horizon of the dataset. The location has higher humidity levels due to its proximity to large bodies of water. There are extreme cases of being less than 60%, which means that air has a lower moisture content and can potentially be considered as dry or less humid.

In [36]:
fig, ax = plt.subplots(figsize=(12, 5))
sns.histplot(df_eda[['pressure']], x='pressure', bins=50, legend=None,
             color='red', kde=True)

plt.ylabel("Count")
plt.xlabel("Pressure (kPa)")
plt.title('Histogram of Pressure of Residential Buildings',
          fontsize=20, pad=20)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Histogram_Pressure.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Histogram_Pressure.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The histogram of the atmospheric pressure indicates that the usual pressure in the location of Kardo is around 101 to 102 kPa. This suggests that the location is most likely in lower terrain than high altitude areas. The variations in the pressure are connected to the changes in the area's temperature.

<h2 style="color:#FFC40F">Additional Time Series Forecasting</h2>

***

We used two other models to compare with the results NBEATSx: Random Forest and Prophet.

Prophet is a time series forecasting model developed by Meta. Using an additive model, it handles trends, seasonality, and holidays in time series data. It has flexible trend modeling, handles various types of seasonality, incorporates holiday effects, and automatically detects change points. Prophet is user-friendly, scalable, and implemented in Python as an open-source library.

Random Forest is a machine learning algorithm commonly used for time series analysis. It applies an ensemble of decision trees to make predictions based on features derived from historical time series data.

The two models are compared to NBEATSx and the baseline Seasonal Naive mentioned in the post in both short-term and long-term forecast horizons.

In [37]:
# Dataframe for Prophet
df_prop = df_day.reset_index()
df_prop = df_prop.rename(columns={'index': 'ds', 'energy_kWh': 'y'})

<h3 style="color:#808080">Short-Term (7-day ahead) Forecast</h3>

In [38]:
# Prophet imports
from prophet.plot import plot_plotly, plot_components_plotly
from prophet import Prophet
import logging
logging.getLogger('prophet').setLevel(logging.WARNING)
logging.getLogger('cmdstanpy').setLevel(logging.WARNING)

In [39]:
# Splitting the dataset
prop_train = df_prop.iloc[:-hor_7]
prop_test = df_prop.iloc[-hor_7:]

# Forecast using Prophet
m = Prophet(holidays=df_holiday)
m.fit(prop_train)
future = m.make_future_dataframe(periods=hor_7)
forecast_7 = m.predict(future)

# Save to dataframe
df_prop_res_7 = (forecast_7[-hor_7:][['ds', 'yhat']]
                 .set_index('ds').rename(columns={'yhat': 'Prophet'}))

In [40]:
# ML imports
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import (mean_absolute_error,
                             mean_absolute_percentage_error)

In [41]:
# Create a features of shifted values
ts_7 = pd.DataFrame()
ts_7['y'] = df_all[['energy_kWh']]

# Lookback window size
window_size = hor_7 * 4

# Create new columns of different lookback period
for w in range(window_size):
    ts_7['y_' + str(w + 1)] = df_all[['energy_kWh']].shift(w + 1)

# Create x dataframe
df_rf_7 = pd.concat([ts_7[window_size:], df_all[window_size:]], axis=1)
df_x_7 = df_rf_7.drop(columns=['energy_kWh', 'y'])

# Create y dataframe
df_y_7 = pd.concat([df_all['energy_kWh'].shift(-i)
                   for i in range(hor_7)], axis=1).dropna()
df_y_7.columns = [f'y{i+1}' for i in range(hor_7)]

In [42]:
# Initialize contrainer of predictions
y_preds_rf = []

# Define x_train/test
x_train = df_x_7[:-hor_7].to_numpy()
x_test = df_x_7[-hor_7:-hor_7+1].to_numpy()

# Define y_train/test
for col in df_y_7.columns:
    y_train = df_y_7[[col]][hor_7*4:-1].to_numpy()
    y_test = df_y_7[[col]][-1:].to_numpy()

    # Fit Scaler on the training set
    scaler_x = StandardScaler().fit(x_train)
    scaler_y = StandardScaler().fit(y_train)

    # Z-score normalization
    x_train = scaler_x.transform(x_train)
    y_train = scaler_y.transform(y_train)
    x_test = scaler_x.transform(x_test)
    y_test = scaler_y.transform(y_test)

    # Results
    df_mape_results = pd.DataFrame()
    param_list = []
    scores = []

    # Param grid
    n_estimators = [100, 200, 300]
    max_depth = [3, 5, 10, 20]
    grid = list(itertools.product(n_estimators, max_depth))

    # Grid search
    for param in tqdm(grid):
        params = {
            'n_estimators': param[0],
            'max_depth': param[1],
        }

        model = RandomForestRegressor(**params, random_state=143)

        # Fit model
        model.fit(x_train, y_train.ravel())

        # Evaluate
        y_pred = model.predict(x_test)
        y_pred = scaler_y.inverse_transform(y_pred[..., np.newaxis])
        y_true = scaler_y.inverse_transform(y_test)
        y_true = y_test.copy()

        # Store results
        param_list.append(str(params))
        scores.append(mean_absolute_percentage_error(y_true, y_pred))

    # Compile results
    df_mape_results['Params'] = param_list
    df_mape_results['Scores'] = scores
    df_mape_results.sort_values(by='Scores', inplace=True)

    # Get model the best model
    best_params = eval(df_mape_results.iloc[0, 0])
    model = RandomForestRegressor(**best_params)
    model.fit(x_train, y_train.ravel())
    y_pred_rf = model.predict(x_test).flatten()
    y_pred_rf = scaler_y.inverse_transform(y_pred_rf[..., np.newaxis])
    y_preds_rf.append(y_pred_rf[0][0])

# Create dataframe of results
df_rf_res_7 = pd.DataFrame(y_preds_rf,
                           index=df_snaive_7.index, columns=['Random_Forest'])

100%|██████████| 12/12 [00:36<00:00,  3.01s/it]


In [43]:
# Combine all predictions and actual
df_predict_7 = df_day.copy()
df_predict_7 = df_predict_7.rename(columns={'energy_kWh': 'Actual'})
df_predict_7 = df_predict_7[-hor_7:]
df_predict_7 = pd.concat([df_predict_7, df_snaive_7, df_rf_res_7,
                          df_prop_res_7, df_nbtx_res_7], axis=1)

# Append the model score to the results dataframe
model_ls = []
mae_ls = []
mape_ls = []
for col in df_predict_7.columns[1:]:
    model_ls.append(col)
    mae_ls.append(mean_absolute_error(df_predict_7['Actual'],
                                      df_predict_7[col]))
    mape_ls.append(mean_absolute_percentage_error(df_predict_7['Actual'],
                                                  df_predict_7[col]))

df_results_7 = pd.DataFrame(
    {'Model': model_ls, 'MAE': mae_ls, 'MAPE': mape_ls})
display(df_results_7)

Model       MAE      MAPE
0  Seasonal_Naive  3.041286  0.112502
1   Random_Forest  2.216157  0.079194
2         Prophet  1.365396  0.052350
3         NBEATSx  1.247066  0.045908

In [44]:
# Plot Results of Different Models
fig, ax = plt.subplots(figsize=(8, 3))
ax.barh(df_results_7['Model'], df_results_7['MAPE'], color='#FAA01D')
ax.set_title('MAPE Comparison of Different Models')
ax.set_xlabel('Mean Absolute Percentage Error')
ax.set_yticklabels(['Seasonal Naive', 'Random Forest', 'Prophet', 'NBEATSx'])
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Short_MAPE_2.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Short_MAPE_2.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In short-term forecasting, NBEATSx beats all other models, including seasonal naïve, random forest, and prophet, by having a MAPE of 4.59%.

In [45]:
# Time Series Prediction
fig, ax = plt.subplots(figsize=(15, 5))
ax.plot(df_predict_7.index, df_predict_7['Actual'],
        label='Actual', color='cornflowerblue', linewidth=3)
ax.plot(df_predict_7.index, df_predict_7['Random_Forest'],
        label='Random Forest', linestyle='--', color='indianred', linewidth=3)
ax.plot(df_predict_7.index, df_predict_7['Prophet'],
        label='Prophet', linestyle='--', color='#FAA01D', linewidth=3)
ax.set_title('Actual vs. Random Forest & Prophet Short-Term Prediction')
ax.set_ylabel('Daily Energy Consumption (kWh)')
ax.set_xlabel('Test Prediction Dates')
ax.legend()
ax.set_ylim(20, 35)
ax.tick_params(axis='x', rotation=15)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Short_Prediction_2.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Short_Prediction_2.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

One noticeable difference between the two forecasts is that prophet predicts higher values than the random forest. Another observation is that rise and fall in the prediction behavior of the prophet better mimics those of the actual values. This might be why the prophet has a lower MAPE between the two.

<h3 style="color:#808080">Long-Term (30-day ahead) Forecast</h3>

In [46]:
# Splitting the dataset
prop_train = df_prop.iloc[:-hor_30]
prop_test = df_prop.iloc[-hor_30:]

# Forecast using Prophet
m = Prophet(holidays=df_holiday)
m.fit(prop_train)
future = m.make_future_dataframe(periods=hor_30)
forecast_30 = m.predict(future)

# Save to dataframe
df_prop_res_30 = (forecast_30[-hor_30:][['ds', 'yhat']]
                  .set_index('ds').rename(columns={'yhat': 'Prophet'}))

In [47]:
# Create a features of shifted values
ts_30 = pd.DataFrame()
ts_30['y'] = df_all[['energy_kWh']]

# Lookback window size
window_size = hor_30

# Create new columns of different lookback period
for w in range(window_size):
    ts_30['y_' + str(w + 1)] = df_all[['energy_kWh']].shift(w + 1)

# Create x dataframe
df_rf_30 = pd.concat([ts_30[window_size:], df_all[window_size:]], axis=1)
df_x_30 = df_rf_30.drop(columns=['energy_kWh', 'y'])

# Create y dataframe
df_y_30 = pd.concat([df_all['energy_kWh'].shift(-i) for i in range(hor_30)],
                    axis=1).dropna()
df_y_30.columns = [f'y{i+1}' for i in range(hor_30)]

In [48]:
# Initialize contrainer of predictions
y_preds_rf = []

# Define x_train/test
x_train = df_x_30[:-hor_30].to_numpy()
x_test = df_x_30[-hor_30:-hor_30+1].to_numpy()

# Define y_train/test
for col in df_y_30.columns:
    y_train = df_y_30[[col]][hor_30:-1].to_numpy()
    y_test = df_y_30[[col]][-1:].to_numpy()

    # Fit Scaler on the training set
    scaler_x = StandardScaler().fit(x_train)
    scaler_y = StandardScaler().fit(y_train)

    # Z-score normalization
    x_train = scaler_x.transform(x_train)
    y_train = scaler_y.transform(y_train)
    x_test = scaler_x.transform(x_test)
    y_test = scaler_y.transform(y_test)

    # Results
    df_mape_results = pd.DataFrame()
    param_list = []
    scores = []

    # Param grid
    n_estimators = [100, 200, 300]
    max_depth = [3, 5, 10, 20]
    grid = list(itertools.product(n_estimators, max_depth))

    # Grid search
    for param in tqdm(grid):
        params = {
            'n_estimators': param[0],
            'max_depth': param[1],
        }

        model = RandomForestRegressor(**params, random_state=143)

        # Fit model
        model.fit(x_train, y_train.ravel())

        # Evaluate
        y_pred = model.predict(x_test)
        y_pred = scaler_y.inverse_transform(y_pred[..., np.newaxis])
        y_true = scaler_y.inverse_transform(y_test)
        y_true = y_test.copy()

        # Store results
        param_list.append(str(params))
        scores.append(mean_absolute_percentage_error(y_true, y_pred))

    # Compile results
    df_mape_results['Params'] = param_list
    df_mape_results['Scores'] = scores
    df_mape_results.sort_values(by='Scores', inplace=True)

    # Get model the best model
    best_params = eval(df_mape_results.iloc[0, 0])
    model = RandomForestRegressor(**best_params)
    model.fit(x_train, y_train.ravel())
    y_pred_rf = model.predict(x_test).flatten()
    y_pred_rf = scaler_y.inverse_transform(y_pred_rf[..., np.newaxis])
    y_preds_rf.append(y_pred_rf[0][0])

# Create dataframe of results
df_rf_res_30 = pd.DataFrame(y_preds_rf,
                            index=df_snaive_30.index,
                            columns=['Random_Forest'])

100%|██████████| 12/12 [00:37<00:00,  3.10s/it]


In [49]:
# Combine all predictions and actual
df_predict_30 = df_day.copy()
df_predict_30 = df_predict_30.rename(columns={'energy_kWh': 'Actual'})
df_predict_30 = df_predict_30[-hor_30:]
df_predict_30 = pd.concat([df_predict_30, df_snaive_30, df_rf_res_30,
                           df_prop_res_30, df_nbtx_res_30], axis=1)

# Append the model score to the results dataframe
model_ls = []
mae_ls = []
mape_ls = []
for col in df_predict_30.columns[1:]:
    model_ls.append(col)
    mae_ls.append(mean_absolute_error(df_predict_30['Actual'],
                                      df_predict_30[col]))
    mape_ls.append(mean_absolute_percentage_error(df_predict_30['Actual'],
                                                  df_predict_30[col]))

df_results_30 = pd.DataFrame(
    {'Model': model_ls, 'MAE': mae_ls, 'MAPE': mape_ls})
display(df_results_30)

Model       MAE      MAPE
0  Seasonal_Naive  4.499367  0.159477
1   Random_Forest  3.430852  0.112313
2         Prophet  2.754180  0.090177
3         NBEATSx  2.562424  0.088149

In [50]:
# Plot Results of Different Models
fig, ax = plt.subplots(figsize=(8, 3))
ax.barh(df_results_30['Model'], df_results_30['MAPE'], color='#FAA01D')
ax.set_title('MAPE Comparison of Different Models')
ax.set_xlabel('Mean Absolute Percentage Error')
ax.set_yticklabels(['Seasonal Naive', 'Random Forest', 'Prophet', 'NBEATSx'])
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Long_MAPE_2.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Long_MAPE_2.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:60%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In long-term forecasting, NBEATSx still beats all other models with a MAPE of 8.81%. As we increase the forecast horizon, it is expected that the accuracy of the forecasting model will drop. Compared to other models in this long-term forecast horizon, it is still one of the best.

In [51]:
# Time Series Prediction
fig, ax = plt.subplots(figsize=(15, 5))
ax.plot(df_predict_30.index, df_predict_30['Actual'],
        label='Actual', color='cornflowerblue', linewidth=3)
ax.plot(df_predict_30.index, df_predict_30['Random_Forest'],
        label='Random Forest', linestyle='--', color='indianred', linewidth=3)
ax.plot(df_predict_30.index, df_predict_30['Prophet'],
        label='Prophet', linestyle='--', color='#FAA01D', linewidth=3)
ax.set_title('Actual vs. Random Forest & Prophet Long-Term Prediction')
ax.set_ylabel('Daily Energy Consumption (kWh)')
ax.set_xlabel('Test Prediction Dates')
ax.legend()
ax.set_ylim(20, 40)
ax.tick_params(axis='x', rotation=15)
plt.tight_layout()

# Save the plot as a PNG file
plt.savefig("./images/Long_Prediction_2.png")

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Long_Prediction_2.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

Compared to the short-term forecast, the prediction of random forest and prophet now intersects. But still, prophets closely follow the up and down of those of the actual value, making it have lower MAPE.

<h2 style="color:#FFC40F">Additional XAI</h2>

***

The code for SHAP mentioned in the main body is indicated below. Random Forest was the model used when applying SHAP to provide explainability. To further provide the value of an explainability model such as SHAP, we contextualize the results by showing how much energy consumption changes when the temperature increases or decreases.

In [52]:
# XAI imports
import shap
from dice_ml import Model, Dice, Data

In [53]:
# Set horizon for XAI
horizon = 120

In [54]:
# Create a features of shifted values
ts = pd.DataFrame()
ts['y'] = df_all[['energy_kWh']]

# Lookback window size
window_size = 7

# Create new columns of different lookback period
for w in range(window_size):
    ts['y_' + str(w + 1)] = df_all[['energy_kWh']].shift(w + 1)

# Adjust dataframe to remove NaN
df_rf = pd.concat([ts[window_size:], df_all[window_size:]], axis=1)
df_rf = df_rf.drop(columns=['energy_kWh'])

In [55]:
# Train-Test split
df_rf_train = df_rf[:-horizon]
df_rf_test = df_rf[-horizon:]
x_train = df_rf_train.drop(['y'], axis=1)
y_train = df_rf_train[['y']]
x_test = df_rf_test.drop(['y'], axis=1)
y_test = df_rf_test[['y']]

# Results
df_mae_results = pd.DataFrame()
param_list = []
scores = []

# Param grid
n_estimators = [100, 200, 300]
max_depth = [3, 5, 10, 20]
grid = list(itertools.product(n_estimators, max_depth))

# 1 day ahead forecasting
for param in tqdm(grid):
    params = {
        'n_estimators': param[0],
        'max_depth': param[1],
    }

    model = RandomForestRegressor(**params, random_state=143)

    # Fit model
    model.fit(x_train, y_train)

    # Evaluate
    y_pred = model.predict(x_test)
    y_true = y_test.copy()

    # Store results
    param_list.append(str(params))
    scores.append(mean_absolute_error(y_true, y_pred))

# Compile results
df_mae_results['Params'] = param_list
df_mae_results['Scores'] = scores
df_mae_results.sort_values(by='Scores', inplace=True)

# Get model the best model
best_params = eval(df_mae_results.iloc[0, 0])
model = RandomForestRegressor(**best_params)
model.fit(x_train, y_train)
y_pred_rf = model.predict(x_test).flatten()

100%|██████████| 12/12 [00:14<00:00,  1.17s/it]


In [56]:
shap_explainer = shap.Explainer(
    model.predict, x_test, feature_names=x_test.columns)
shap_values = shap_explainer(x_test.iloc[:, :])
shap_explanation = shap.Explanation(shap_values.values[:, :],
                                    shap_values.base_values[0],
                                    shap_values.data,
                                    feature_names=x_test.columns)

Permutation explainer: 121it [02:01,  1.08s/it]                         


In [57]:
# Global feature importance
# shap.summary_plot(shap_explanation, plot_type='bar', plot_size=(14, 10))

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/SHAP.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

The top feature in predicting energy consumption is temperature, humidity, and previous lag value. As the descriptive analysis corroborates, energy consumption decreases as temperature increases. We will explain a single forecast on our test set to further investigate this generalization.

In [58]:
# Waterfall diagram
# shap.plots.waterfall(shap_explanation[4])

# Create an HTML img tag to display the image
img_tag = (f'<img src="./images/Waterfall.png" alt="plots" style='
           '"display:block;margin-left:auto;margin-right:auto;width:80%;">')

# Display the img tag in the Jupyter Notebook
display(HTML(img_tag))
plt.close()

In this plot, we can see coming from the global average of energy consumption, the high-temperature value (relative to other time of the year) push the consumption towards lower energy consumption. Let us determine how big of a change in energy consumption each temperature increase.

In [59]:
model1 = Model(model=model, backend='sklearn', model_type='regressor')
dice_data = Data(
    dataframe=df_rf_test,
    continuous_features=['y_1', 'y_2', 'y_3', 'y_4', 'y_5', 'y_6', 'y_7',
                         'temperature', 'humidity', 'pressure'],
    outcome_name='y'
)
dice_exp = Dice(dice_data, model1, method='genetic')

In [60]:
df_rf_test_x = df_rf_test.drop(columns='y')

In [61]:
power_deltas = []
for i in range(1, 16):
    np.random.seed(630)
    cfes1 = dice_exp.generate_counterfactuals(
        df_rf_test_x.iloc[[i]],
        total_CFs=1,
        desired_range=[25, 26],
        features_to_vary=['temperature'])
    new_power = dice_exp.final_cfs_df['y'][0]
    new_temp = dice_exp.final_cfs_df['temperature'][0]
    old_power = model.predict(df_rf_test_x.iloc[[i]])[0]
    old_temp = df_rf_test_x.iloc[[i]]['temperature'][0]
    power_delta = ((new_power - old_power) / (new_temp - old_temp))
    power_deltas.append(power_delta)

100%|██████████| 1/1 [00:01<00:00,  1.14s/it]


In [62]:
print(f'{np.mean(power_deltas):.3f} kWh change in energy consumption change for every 1 degree '
      'Celcius increase in temperature')

-0.289 kWh change in energy consumption change for every 1 degree Celcius increase in temperature
